File: androidApp/feature-history/src/main/kotlin/dev/roozbahani/trailmetrics/feature/history/HistoryViewModel.kt
package dev.roozbahani.trailmetrics.feature.history

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dev.roozbahani.trailmetrics.domain.model.ActivityRecord
import dev.roozbahani.trailmetrics.domain.repository.ActivityHistoryRepository
import dev.roozbahani.trailmetrics.feature.history.util.deleteSnapshotFile
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class HistoryViewModel(
    private val activityHistoryRepository: ActivityHistoryRepository
) : ViewModel() {

    val uiState: StateFlow<HistoryUiState> = activityHistoryRepository.observeActivities()
        .map { activities -> HistoryUiState(activities, false) }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5_000L),
            initialValue = HistoryUiState(isLoading = true)
        )

    fun onDeleteActivity(activity: ActivityRecord) {
        viewModelScope.launch {
            activityHistoryRepository.deleteActivity(activity.id)
            deleteSnapshotFile(activity.snapshotFilePath)
        }
    }
}

data class HistoryUiState(
    val activities: List<ActivityRecord> = emptyList(),
    val isLoading: Boolean = false
) {
    val isEmpty: Boolean
        get() = activities.isEmpty() && !isLoading
}

File: iosApp/Packages/History/Sources/History/HistoryViewModel.swift
//
//  HistoryViewModel.swift
//  History
//
//  Created by Ali Roozbahani on 09.09.26.
//

import Foundation
import SharedKit

@MainActor
public class HistoryViewModel: ObservableObject {
    @Published public var activities: [ActivityRecord] = []
    @Published public var isLoading = true

    public var isEmpty: Bool {
        activities.isEmpty && !isLoading
    }

    private let activityHistoryRepository: ActivityHistoryRepository

    public init(activityHistoryRepository: ActivityHistoryRepository = KoinHelper().getActivityHistoryRepository()) {
        self.activityHistoryRepository = activityHistoryRepository
    }

    public func observe() async {
        for await activities in activityHistoryRepository.observeActivities() {
            self.activities = activities
            self.isLoading = false
        }
    }

    public func onDeleteActivity(_ activity: ActivityRecord) {
        Task {
            try? await activityHistoryRepository.deleteActivity(id: activity.id)
            deleteSnapshotFile(forStoredPath: activity.snapshotFilePath)
        }
    }
}
